jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/StandardDescriptorProcessor.java

```diff
@@ -1028,69 +1038,134 @@ public void addWelcomeFiles(WebAppContext context, XmlParser.Node node, Descript
+    {
+        Objects.requireNonNull(servletName);
+        Objects.requireNonNull(path);
+        Objects.requireNonNull(context);
+        Objects.requireNonNull(descriptor);
+
+        ListIterator<ServletMapping> listItor = _servletMappings.listIterator();
+        boolean found = false;
+        while (listItor.hasNext() && !found)
+        {
+            ServletMapping originalMapping = listItor.next();
+            String[] originalPathSpecs = originalMapping.getPathSpecs();
+            if (originalPathSpecs == null || (originalPathSpecs.length == 0))
+                continue;
+
+            for (String originalPath : originalPathSpecs)
+            {
+                if (!originalPath.equals(path))
+                    continue;
+
+                if (servletName.equals(originalMapping.getServletName()))
+                {
+                   //Same path mapped to same servlet, ignore
+                   return false;
+               }
+               else
+               {
+                   //Same path mapped to different servlet
+                   switch (originalMapping.getSource().getOrigin())
+                   {
+                       case EMBEDDED ->
+                       {
+                           if (descriptor instanceof DefaultsDescriptor)
+                               return false;  //webdefault.xml cannot override a path mapping
+
+                           //Remove existing mapping
+                           removeMapping(originalMapping, path);
+                           //Tidy up by removing the originMapping if it now has no paths
+                           if (originalMapping.getPathSpecs() == null || originalMapping.getPathSpecs().length == 0)
+                               listItor.remove();
+                           return true;
+                       }
+                       case DESCRIPTOR ->
+                       {
+                           //if the descriptor of the original mapping and the new mapping are the same, that is an error
+                           if (originalMapping.getSource().getResource().equals(descriptor.getResource()))
+                               throw new IllegalStateException("Duplicate mappings for " + path);
+
+                           //if the original mapping came from the defaults descriptor it can be overridden by any other descriptor
+                           if (originalMapping.isFromDefaultDescriptor())
+                           {
+                               //Remove existing mapping
+                               removeMapping(originalMapping, path);
+                               //Tidy up by removing the originMapping if it now has no paths
+                               if (originalMapping.getPathSpecs() == null || originalMapping.getPathSpecs().length == 0)
+                                   listItor.remove();
+                               return true;
```

Comment:
How can we return true here without checking other paths?

Reply:
Oh we can return true here (or false below) because we know there are no duplicate paths in the mappings.   If we find one match, then it will be the one and only match!  correct?

Reply:
Pushed an alternate implementation based on this understanding